src/dmd/escape.d

```diff
@@ -585,9 +585,142 @@ bool checkThrowEscape(Scope* sc, Expression e, bool gag)
+                {
+                    if (!gag)
+                        error(e.loc, "scope variable `%s` may not be copied into allocated memory", v.toChars());
+                    result = true;
+                }
+                continue;
+            }
+        }
+        else if (v.storage_class & STC.variadic && p == sc.func)
+        {
+            Type tb = v.type.toBasetype();
+            if (tb.ty == Tarray || tb.ty == Tsarray)
+            {
+                if (!gag)
+                    error(e.loc, "copying `%s` into allocated memory escapes a reference to variadic parameter `%s`", e.toChars(), v.toChars());
+                result = false;
+            }
+        }
+        else
+        {
+            //printf("no infer for %s in %s, %d\n", v.toChars(), sc.func.ident.toChars(), __LINE__);
+            v.doNotInferScope = true;
+        }
+    }
+
+    foreach (VarDeclaration v; er.byref)
+    {
+        if (log) printf("byref `%s`\n", v.toChars());
+
+        void escapingRef(VarDeclaration v)
+        {
+            if (!gag)
+            {
+                const(char)* kind = (v.storage_class & STC.parameter) ? "parameter" : "local";
+                error(e.loc, "copying `%s` into allocated memory escapes a reference to %s variable `%s`",
+                    e.toChars(), kind, v.toChars());
+            }
+            result = true;
+        }
+
+        if (v.isDataseg())
+            continue;
+
+        Dsymbol p = v.toParent2();
+
+        if ((v.storage_class & (STC.ref_ | STC.out_)) == 0)
+        {
+            if (p == sc.func)
+            {
+                escapingRef(v);
+                continue;
+            }
+        }
+
+        /* Check for returning a ref variable by 'ref', but should be 'return ref'
+         * Infer the addition of 'return', or set result to be the offending expression.
+         */
+        if (v.storage_class & (STC.ref_ | STC.out_))
+        {
+            if (global.params.useDIP25 &&
+                     sc._module && sc._module.isRoot())
+            {
+                // Only look for errors if in module listed on command line
+
+                if (p == sc.func)
+                {
+                    //printf("escaping reference to local ref variable %s\n", v.toChars());
+                    //printf("storage class = x%llx\n", v.storage_class);
+                    escapingRef(v);
+                    continue;
+                }
+                // Don't need to be concerned if v's parent does not return a ref
+                FuncDeclaration fd = p.isFuncDeclaration();
+                if (fd && fd.type && fd.type.ty == Tfunction)
+                {
+                    TypeFunction tf = cast(TypeFunction)fd.type;
+                    if (tf.isref)
+                    {
+                        if (!gag)
+                            error(e.loc, "storing reference to outer local variable `%s` into allocated memory", v.toChars());
```

Comment:
Not sure I understand what the intention is, but if my understanding is correct I suggest:

> storing a reference to outer local variable `%s` into GC-allocated memory causes it to escape